src/notebooks/controllers/kernelSource/kernelSourceCommandHandler.ts

```diff
@@ -76,16 +74,47 @@ export class KernelSourceCommandHandler implements IExtensionSyncActivationServi
                                 label: DataScience.localPythonEnvironments(),
                                 detail: DataScience.pickLocalKernelPythonEnvTitle(),
                                 command: 'jupyter.kernel.selectLocalPythonEnvironment'
-                            },
-                            {
-                                label: DataScience.localKernelSpecs(),
-                                detail: DataScience.pickLocalKernelSpecTitle(),
-                                command: 'jupyter.kernel.selectLocalKernelSpec'
                             }
                         ];
                     }
                 })
             );
+            const registerKernelSpecsSource = () => {
+                if (this.kernelSpecsSourceRegistered) {
+                    return;
+                }
+                if (this.kernelFinder.kernels.some((k) => k.kind === 'startUsingLocalKernelSpec')) {
+                    this.kernelSpecsSourceRegistered = true;
+                    this.localDisposables.push(
+                        notebooks.registerKernelSourceActionProvider(JupyterNotebookView, {
+                            provideNotebookKernelSourceActions: () => {
+                                return [
+                                    {
+                                        label: DataScience.localKernelSpecs(),
+                                        detail: DataScience.pickLocalKernelSpecTitle(),
+                                        command: 'jupyter.kernel.selectLocalKernelSpec'
+                                    }
+                                ];
+                            }
+                        })
+                    );
+                    this.localDisposables.push(
+                        notebooks.registerKernelSourceActionProvider(InteractiveWindowView, {
+                            provideNotebookKernelSourceActions: () => {
+                                return [
+                                    {
+                                        label: DataScience.localKernelSpecs(),
+                                        detail: DataScience.pickLocalKernelSpecTitle(),
+                                        command: 'jupyter.kernel.selectLocalKernelSpec'
+                                    }
+                                ];
+                            }
+                        })
+                    );
+                }
+            };
+            registerKernelSpecsSource();
+            this.kernelFinder.onDidChangeKernels(() => registerKernelSpecsSource(), this, this.localDisposables);
```

Comment:
Does this need to be added to `localDisposables`?

Reply:
Yes, we want to be able to dispose these later in line 42.